File: packages/tls-fingerprint/ios/Sources/TLSFingerprintPlugin/TLSFingerprintImpl.swift
import Foundation

/**
 Native iOS implementation for the TLSFingerprint plugin.

 Responsibilities:
 - Perform platform-specific operations
 - Throw typed TLSFingerprintError values on failure

 Forbidden:
 - Accessing CAPPluginCall
 - Referencing Capacitor APIs
 - Constructing JS payloads
 */
@objc public final class TLSFingerprintImpl: NSObject {

    // MARK: - Constants

    private static let timeoutSeconds: TimeInterval = 10

    // MARK: - Properties

    /**
     Static plugin configuration.

     Injected once during plugin initialization.
     Must not be mutated after being set.
     */
    private var config: TLSFingerprintConfig?

    // Initializer
    override init() {
        super.init()
    }

    // MARK: - Configuration

    /**
     Applies static plugin configuration.

     This method MUST be called exactly once
     from the Plugin layer during `load()`.

     Responsibilities:
     - Store immutable configuration
     - Configure runtime logging behavior
     */
    func applyConfig(_ config: TLSFingerprintConfig) {
        precondition(
            self.config == nil,
            "TLSFingerprintImpl.applyConfig(_:) must be called exactly once"
        )
        self.config = config
        TLSFingerprintLogger.verbose = config.verboseLogging

        TLSFingerprintLogger.debug(
            "Configuration applied. Verbose logging:",
            config.verboseLogging
        )
    }

    // MARK: - Single fingerprint

    /**
     Validates the SSL certificate of a HTTPS endpoint
     using a single SHA-256 fingerprint.

     Resolution order:
     1. Runtime fingerprint argument
     2. Static configuration fingerprint

     - Throws: `TLSFingerprintError.unavailable`
     if no fingerprint is available.
     */
    func checkCertificate(
        urlString: String,
        fingerprintFromArgs: String?
    ) async throws -> TLSFingerprintResult {

        let fingerprint =
            fingerprintFromArgs ??
            config?.fingerprint

        guard let expectedFingerprint = fingerprint else {
            throw TLSFingerprintError.unavailable(
                TLSFingerprintErrorMessages.noFingerprintsProvided
            )
        }

        return try await performCheck(
            urlString: urlString,
            fingerprints: [expectedFingerprint]
        )
    }

    // MARK: - Multiple fingerprints

    /**
     Validates the SSL certificate of a HTTPS endpoint
     using multiple allowed SHA-256 fingerprints.

     A match is considered valid if ANY provided
     fingerprint matches the server certificate.

     - Throws: `TLSFingerprintError.unavailable`
     if no fingerprints are available.
     */
    func checkCertificates(
        urlString: String,
        fingerprintsFromArgs: [String]?
    ) async throws -> TLSFingerprintResult {

        let fingerprints =
            fingerprintsFromArgs ??
            config?.fingerprints

        guard let fps = fingerprints, !fps.isEmpty else {
            throw TLSFingerprintError.unavailable(
                TLSFingerprintErrorMessages.noFingerprintsProvided
            )
        }

        return try await performCheck(
            urlString: urlString,
            fingerprints: fps
        )
    }

    // MARK: - Shared implementation

    /**
     Performs SSL fingerprint validation for a given HTTPS URL.

     Evaluation order:

     1. Excluded domains → bypass validation entirely.
     2. Fingerprint-based validation.

     This method uses async/await and wraps the URLSession
     delegate flow inside a continuation.

     - Throws:
     - `TLSFingerprintError.unknownType`
     - `TLSFingerprintError.initFailed`
     */
    private func performCheck(
        urlString: String,
        fingerprints: [String]
    ) async throws -> TLSFingerprintResult {

        guard let url = TLSFingerprintUtils.httpsURL(from: urlString) else {
            throw TLSFingerprintError.unknownType(
                TLSFingerprintErrorMessages.invalidUrlMustBeHttps
            )
        }

        let host = url.host?.lowercased() ?? ""

        // -----------------------------------------------------------------------
        // EXCLUDED DOMAIN MODE
        // -----------------------------------------------------------------------

        /**
         If the request host matches an excluded domain,
         we still perform the connection to retrieve the actual fingerprint
         for parity with Android behavior.

         The connection uses a permissive trust manager.
         System trust chain is NOT explicitly evaluated.
         */
        if isExcludedDomain(host) {
            TLSFingerprintLogger.debug("TLSFingerprint excluded domain:", host)

            return try await withCheckedThrowingContinuation { continuation in
                let configuration = URLSessionConfiguration.ephemeral
                configuration.timeoutIntervalForRequest = Self.timeoutSeconds
                configuration.timeoutIntervalForResource = Self.timeoutSeconds

                let delegate = TLSFingerprintDelegate(
                    expectedFingerprints: [],
                    excludedDomains: config?.excludedDomains ?? [],
                    completion: { result in
                        continuation.resume(returning: result)
                    },
                    verboseLogging: config?.verboseLogging ?? false
                )

                let session = URLSession(
                    configuration: configuration,
                    delegate: delegate,
                    delegateQueue: nil
                )

                delegate.setSession(session)

                let timeoutWorkItem = DispatchWorkItem { [weak delegate] in
                    guard let delegate = delegate else { return }
                    let shouldTimeout = delegate.trySetCompleted()
                    if shouldTimeout {
                        session.invalidateAndCancel()
                        continuation.resume(returning: TLSFingerprintResult(
                            fingerprintMatched: false,
                            error: TLSFingerprintErrorMessages.timeout,
                            errorCode: "TIMEOUT"
                        ))
                    }
                }

                DispatchQueue.global().asyncAfter(deadline: .now() + Self.timeoutSeconds, execute: timeoutWorkItem)

                session.dataTask(with: url).resume()
            }
        }

        return try await withCheckedThrowingContinuation { continuation in

            let configuration = URLSessionConfiguration.ephemeral
            configuration.timeoutIntervalForRequest = 10
            configuration.timeoutIntervalForResource = 10

            let delegate = TLSFingerprintDelegate(
                expectedFingerprints: fingerprints,
                excludedDomains: config?.excludedDomains ?? [],
                completion: { result in
                    continuation.resume(returning: result)
                },
                verboseLogging: config?.verboseLogging ?? false
            )

            let session = URLSession(
                configuration: configuration,
                delegate: delegate,
                delegateQueue: nil
            )

            delegate.setSession(session)

            let timeoutWorkItem = DispatchWorkItem { [weak delegate] in
                guard let delegate = delegate else { return }
                let shouldTimeout = delegate.trySetCompleted()
                if shouldTimeout {
                    session.invalidateAndCancel()
                    continuation.resume(returning: TLSFingerprintResult(
                        fingerprintMatched: false,
                        error: TLSFingerprintErrorMessages.timeout,
                        errorCode: "TIMEOUT"
                    ))
                }
            }

            DispatchQueue.global().asyncAfter(deadline: .now() + Self.timeoutSeconds, execute: timeoutWorkItem)

            session.dataTask(with: url).resume()
        }
    }

    /**
     Determines whether the given host
     matches one of the configured excluded domains.

     Matching rules:
     - Exact match
     - Subdomain match

     Matching is case-insensitive.
     */
    private func isExcludedDomain(_ host: String) -> Bool {
        let hostLower = host.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        return config?.excludedDomains.contains { excluded in
            let excludedLower = excluded.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
            // Match exact domain or any subdomain (e.g., api.example.com matches example.com)
            return hostLower == excludedLower || hostLower.hasSuffix("." + excludedLower)
        } ?? false
    }
}

File: packages/tls-fingerprint/android/src/main/java/io/capkit/settings/TLSFingerprintImpl.kt
package io.capkit.tlsfingerprint

import android.content.Context
import io.capkit.tlsfingerprint.config.TLSFingerprintConfig
import io.capkit.tlsfingerprint.error.TLSFingerprintError
import io.capkit.tlsfingerprint.error.TLSFingerprintErrorMessages
import io.capkit.tlsfingerprint.logger.TLSFingerprintLogger
import io.capkit.tlsfingerprint.model.TLSFingerprintResultModel
import io.capkit.tlsfingerprint.utils.TLSFingerprintUtils
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.URL
import java.security.cert.Certificate
import java.security.cert.X509Certificate
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

/**
 * Native Android implementation for the TLSFingerprint plugin.
 *
 * Responsibilities:
 * - Perform platform-specific SSL pinning logic
 * - Interact with system networking APIs
 * - Throw typed TLSFingerprintError values on failure
 *
 * Forbidden:
 * - Accessing PluginCall
 * - Referencing Capacitor APIs
 * - Constructing JavaScript payloads
 */
class TLSFingerprintImpl(
  private val context: Context,
) {
  // -----------------------------------------------------------------------------
  // Constants
  // -----------------------------------------------------------------------------

  companion object {
    private const val TIMEOUT_MS = 10_000
  }

  // -----------------------------------------------------------------------------
  // Properties
  // -----------------------------------------------------------------------------

  /**
   * Cached plugin configuration.
   * Injected once during plugin initialization.
   */
  private lateinit var config: TLSFingerprintConfig

  // -----------------------------------------------------------------------------
  // Configuration
  // -----------------------------------------------------------------------------

  /**
   * Applies plugin configuration.
   *
   * This method MUST be called exactly once from the plugin's load() method.
   * It translates static configuration into runtime behavior
   * (e.g. enabling verbose logging).
   */
  fun updateConfig(newConfig: TLSFingerprintConfig) {
    this.config = newConfig
    TLSFingerprintLogger.verbose = newConfig.verboseLogging

    TLSFingerprintLogger.debug(
      "Configuration applied. Verbose logging:",
      newConfig.verboseLogging.toString(),
    )
  }

  // -----------------------------------------------------------------------------
  // Single fingerprint
  // -----------------------------------------------------------------------------

  /**
   * Validates the SSL certificate of a HTTPS endpoint
   * using a single SHA-256 fingerprint.
   *
   * Resolution order:
   * 1. Runtime fingerprint argument
   * 2. Static configuration fingerprint
   *
   * @throws TLSFingerprintError.Unavailable if no fingerprint available.
   */
  @Throws(TLSFingerprintError::class)
  fun checkCertificate(
    urlString: String,
    fingerprintFromArgs: String?,
  ): TLSFingerprintResultModel {
    val fingerprint =
      fingerprintFromArgs ?: config.fingerprint

    if (fingerprint != null) {
      return performCheck(
        urlString = urlString,
        fingerprints = listOf(fingerprint),
      )
    }

    throw TLSFingerprintError.Unavailable(
      TLSFingerprintErrorMessages.NO_FINGERPRINTS_PROVIDED,
    )
  }

  // -----------------------------------------------------------------------------
  // Multiple fingerprints
  // -----------------------------------------------------------------------------

  /**
   * Validates the SSL certificate of a HTTPS endpoint
   * using multiple allowed SHA-256 fingerprints.
   *
   * A match is considered valid if ANY provided fingerprint matches.
   *
   * @throws TLSFingerprintError.Unavailable if no fingerprints available.
   */
  @Throws(TLSFingerprintError::class)
  fun checkCertificates(
    urlString: String,
    fingerprintsFromArgs: List<String>?,
  ): TLSFingerprintResultModel {
    val fingerprints =
      fingerprintsFromArgs?.takeIf { it.isNotEmpty() }
        ?: config.fingerprints.takeIf { it.isNotEmpty() }

    if (fingerprints != null) {
      return performCheck(
        urlString = urlString,
        fingerprints = fingerprints,
      )
    }

    throw TLSFingerprintError.Unavailable(
      TLSFingerprintErrorMessages.NO_FINGERPRINTS_PROVIDED,
    )
  }

  // -----------------------------------------------------------------------------
  // Shared implementation
  // -----------------------------------------------------------------------------

  /**
   * Determines and executes the SSL fingerprint validation for the given request.
   *
   * Evaluation order:
   *
   * 1. Excluded domain → bypass validation entirely.
   * 2. Fingerprint-based validation.
   *
   * @throws TLSFingerprintError when validation fails.
   */
  @Throws(TLSFingerprintError::class)
  private fun performCheck(
    urlString: String,
    fingerprints: List<String>,
  ): TLSFingerprintResultModel {
    val url =
      TLSFingerprintUtils.httpsUrl(urlString)
        ?: throw TLSFingerprintError.UnknownType(
          TLSFingerprintErrorMessages.INVALID_URL_MUST_BE_HTTPS,
        )

    val host = url.host

    // -----------------------------------------------------------------------------
    // EXCLUDED DOMAIN MODE
    // -----------------------------------------------------------------------------

    /**
     * If the request host matches an excluded domain,
     * fingerprint validation is bypassed.
     * The connection uses a permissive TrustManager.
     * System trust chain is NOT explicitly evaluated.
     *
     * Matching rules:
     * - Exact match
     * - Subdomain match
     */
    if (config.excludedDomains.any { excluded ->
        val excludedLower = excluded.lowercase().trim()
        val hostLower = host.lowercase().trim()
        // Match exact domain or any subdomain (e.g., api.example.com matches example.com)
        hostLower == excludedLower || hostLower.endsWith(".$excludedLower")
      }
    ) {
      TLSFingerprintLogger.debug("TLSFingerprint excluded domain:", host)

      try {
        val certificate = getCertificate(url)
        val actualFingerprint =
          TLSFingerprintUtils.normalizeFingerprint(
            TLSFingerprintUtils.sha256Fingerprint(certificate),
          )

        return TLSFingerprintResultModel(
          actualFingerprint = actualFingerprint,
          fingerprintMatched = true,
          excludedDomain = true,
          mode = "excluded",
          errorCode = "EXCLUDED_DOMAIN",
          error = TLSFingerprintErrorMessages.EXCLUDED_DOMAIN,
        )
      } catch (e: SocketTimeoutException) {
        throw TLSFingerprintError.Timeout(TLSFingerprintErrorMessages.TIMEOUT)
      } catch (e: IOException) {
        throw TLSFingerprintError.NetworkError(TLSFingerprintErrorMessages.NETWORK_ERROR)
      }
    }

    // -----------------------------------------------------------------------------
    // FINGERPRINT MODE
    // -----------------------------------------------------------------------------

    /**
     * Fingerprint-based pinning.
     *
     * Only the leaf certificate fingerprint is compared.
     * The system trust chain is NOT evaluated in this mode.
     */
    val certificate: Certificate
    try {
      certificate = getCertificate(url)
    } catch (e: SocketTimeoutException) {
      throw TLSFingerprintError.Timeout(TLSFingerprintErrorMessages.TIMEOUT)
    } catch (e: IOException) {
      throw TLSFingerprintError.NetworkError(TLSFingerprintErrorMessages.NETWORK_ERROR)
    }

    val actualFingerprint =
      TLSFingerprintUtils.normalizeFingerprint(
        TLSFingerprintUtils.sha256Fingerprint(certificate),
      )

    val normalizedExpected =
      fingerprints.map {
        TLSFingerprintUtils.normalizeFingerprint(it)
      }

    val matchedFingerprint =
      normalizedExpected.firstOrNull {
        it == actualFingerprint
      }

    val matched = matchedFingerprint != null

    val errorCode = if (matched) "" else "PINNING_FAILED"
    val errorMessage = if (matched) "" else TLSFingerprintErrorMessages.PINNING_FAILED

    TLSFingerprintLogger.debug(
      "TLSFingerprint matched:",
      matched.toString(),
    )

    return TLSFingerprintResultModel(
      actualFingerprint = actualFingerprint,
      fingerprintMatched = matched,
      matchedFingerprint = matchedFingerprint,
      mode = "fingerprint",
      errorCode = errorCode,
      error = errorMessage,
    )
  }

  // -----------------------------------------------------------------------------
  // Certificate retrieval
  // -----------------------------------------------------------------------------

  /**
   * Opens a TLS connection and extracts
   * the server leaf certificate.
   *
   * SECURITY MODEL:
   * - A permissive TrustManager is used intentionally.
   * - The system trust chain is NOT validated.
   * - Validation is performed manually via fingerprint comparison.
   */
  @Throws(Exception::class)
  private fun getCertificate(url: URL): Certificate {
    val trustManagers =
      arrayOf<TrustManager>(
        object : X509TrustManager {
          override fun getAcceptedIssuers() = emptyArray<X509Certificate>()

          override fun checkClientTrusted(
            certs: Array<X509Certificate>,
            authType: String,
          ) {}

          override fun checkServerTrusted(
            certs: Array<X509Certificate>,
            authType: String,
          ) {}
        },
      )

    val sslContext =
      SSLContext.getInstance("TLS")

    sslContext.init(
      null,
      trustManagers,
      java.security.SecureRandom(),
    )

    val connection =
      url.openConnection() as HttpsURLConnection

    try {
      connection.sslSocketFactory =
        sslContext.socketFactory

      connection.connectTimeout = TIMEOUT_MS
      connection.readTimeout = TIMEOUT_MS

      connection.connect()

      val certificate =
        connection.serverCertificates.first()

      return certificate
    } finally {
      try {
        connection.disconnect()
      } catch (_: Exception) {
        // Ignore disconnect errors
      }
    }
  }
}
